File: src/evolution/self-modify.ts
import { existsSync, copyFileSync, mkdirSync } from "node:fs";
import { resolve, relative } from "node:path";
import { createLogger } from "../logger";

const log = createLogger("evolution:selfmod");

const PROJECT_ROOT = process.cwd();
const BACKUP_DIR = resolve(PROJECT_ROOT, "data", "backups");

const ALLOWED_GLOBS = [
  "user-space/**",
  "src/agent/tools/builtin/**",
  "config/**",
];

const DENIED_PATHS = [
  "src/evolution/",
  "src/config/",
  "src/logger.ts",
  "src/index.ts",
  "src/worker.ts",
  "src/gateway/server.ts",
  "package.json",
  "tsconfig.json",
  ".env",
  ".git/",
];

export interface ModifyResult {
  success: boolean;
  reason?: string;
  backupPath?: string;
}

export class SelfModifier {
  private changeLog: Array<{
    timestamp: string;
    file: string;
    action: string;
    rationale: string;
  }> = [];

  isAllowed(filePath: string): boolean {
    const rel = relative(PROJECT_ROOT, resolve(PROJECT_ROOT, filePath));

    for (const denied of DENIED_PATHS) {
      if (rel.startsWith(denied) || rel === denied.replace(/\/$/, "")) {
        return false;
      }
    }

    for (const pattern of ALLOWED_GLOBS) {
      const prefix = pattern.replace("/**", "");
      if (rel.startsWith(prefix)) return true;
    }

    return false;
  }

  async modify(
    filePath: string,
    content: string,
    rationale: string,
  ): Promise<ModifyResult> {
    const absPath = resolve(PROJECT_ROOT, filePath);
    const rel = relative(PROJECT_ROOT, absPath);

    if (!this.isAllowed(rel)) {
      log.warn("Self-modify blocked", { path: rel, rationale });
      return {
        success: false,
        reason: `Path not in allowlist: ${rel}. Allowed: ${ALLOWED_GLOBS.join(", ")}`,
      };
    }

    // Backup existing file
    let backupPath: string | undefined;
    if (existsSync(absPath)) {
      backupPath = await this.backup(absPath, rel);
    }

    // Ensure parent directory exists
    const parentDir = absPath.substring(0, absPath.lastIndexOf("/"));
    if (!existsSync(parentDir)) {
      mkdirSync(parentDir, { recursive: true });
    }

    // Validate TypeScript syntax for .ts files
    if (filePath.endsWith(".ts")) {
      const valid = await this.validateSyntax(content);
      if (!valid) {
        return {
          success: false,
          reason: "TypeScript syntax validation failed",
          backupPath,
        };
      }
    }

    await Bun.write(absPath, content);

    this.changeLog.push({
      timestamp: new Date().toISOString(),
      file: rel,
      action: existsSync(absPath) ? "modify" : "create",
      rationale,
    });

    log.info("Self-modify applied", { path: rel, rationale });
    return { success: true, backupPath };
  }

  private async backup(absPath: string, rel: string): Promise<string> {
    if (!existsSync(BACKUP_DIR)) {
      mkdirSync(BACKUP_DIR, { recursive: true });
    }

    const timestamp = Date.now();
    const safeName = rel.replace(/\//g, "__");
    const backupPath = resolve(BACKUP_DIR, `${timestamp}_${safeName}`);
    copyFileSync(absPath, backupPath);
    return backupPath;
  }

  private async validateSyntax(content: string): Promise<boolean> {
    try {
      // Use Bun's built-in transpiler for fast syntax checking
      new Bun.Transpiler({ loader: "ts" }).transformSync(content);
      return true;
    } catch (err) {
      log.warn("Syntax validation failed", {
        error: err instanceof Error ? err.message : String(err),
      });
      return false;
    }
  }

  getChangeLog() {
    return [...this.changeLog];
  }
}

let _modifier: SelfModifier | null = null;

export function getSelfModifier(): SelfModifier {
  if (!_modifier) _modifier = new SelfModifier();
  return _modifier;
}

File: src/evolution/self_modify.py
"""Safe self-modification with allowlist/denylist, matching TS evolution/self-modify.ts."""

from __future__ import annotations

import ast
import shutil
import time
from dataclasses import dataclass
from pathlib import Path

from src.utils.logger import create_logger

log = create_logger("evolution:self-modify")

PROJECT_ROOT = Path.cwd()

ALLOWED_GLOBS = [
    "user-space/",
    "src/agent/tools/builtin/",
    "config/",
]

DENIED_PATHS = [
    "src/evolution/",
    "src/config/",
    "src/logger.",
    "src/index.",
    "src/worker.",
    "src/gateway/server.",
    "package.json",
    "tsconfig.json",
    "pyproject.toml",
    ".env",
    ".git/",
]


@dataclass
class ModifyResult:
    success: bool
    reason: str | None = None
    backup_path: str | None = None


class SelfModifier:
    def __init__(self) -> None:
        self._change_log: list[dict] = []

    def is_allowed(self, file_path: str) -> bool:
        resolved = (PROJECT_ROOT / file_path).resolve()
        try:
            rel = str(resolved.relative_to(PROJECT_ROOT))
        except ValueError:
            return False

        for denied in DENIED_PATHS:
            if rel.startswith(denied):
                return False

        for allowed in ALLOWED_GLOBS:
            if rel.startswith(allowed):
                return True

        return False

    async def modify(self, file_path: str, content: str, rationale: str) -> ModifyResult:
        if not self.is_allowed(file_path):
            return ModifyResult(success=False, reason=f"Path not allowed: {file_path}")

        abs_path = (PROJECT_ROOT / file_path).resolve()
        backup_path = None

        if abs_path.exists():
            backup_path = await self._backup(abs_path, file_path)

        abs_path.parent.mkdir(parents=True, exist_ok=True)

        if file_path.endswith(".py"):
            if not self._validate_syntax(content):
                return ModifyResult(success=False, reason="Python syntax validation failed")

        abs_path.write_text(content, encoding="utf-8")

        self._change_log.append({
            "timestamp": time.time(),
            "file": file_path,
            "action": "modify",
            "rationale": rationale,
        })

        log.info("File modified", {"file": file_path, "rationale": rationale})
        return ModifyResult(success=True, backup_path=backup_path)

    async def _backup(self, abs_path: Path, rel: str) -> str:
        backup_dir = PROJECT_ROOT / "data" / "backups"
        backup_dir.mkdir(parents=True, exist_ok=True)
        ts = int(time.time())
        safe_name = rel.replace("/", "_").replace("\\", "_")
        backup_file = backup_dir / f"{ts}_{safe_name}"
        shutil.copy2(str(abs_path), str(backup_file))
        return str(backup_file)

    def _validate_syntax(self, content: str) -> bool:
        try:
            ast.parse(content)
            return True
        except SyntaxError:
            return False

    def get_change_log(self) -> list[dict]:
        return list(self._change_log)


_modifier: SelfModifier | None = None


def get_self_modifier() -> SelfModifier:
    global _modifier
    if _modifier is None:
        _modifier = SelfModifier()
    return _modifier
